pkg/securelaunch/helpers.go

```diff
@@ -0,0 +1,270 @@
+	if !srcFileFound {
+		return "", fmt.Errorf("File to be written doesnt exist:")
+	}
+
+	d, err := ioutil.ReadFile(src)
+	if err != nil {
+		return "", fmt.Errorf("Error reading src file %s, err=%v", src, err)
+	}
+
+	target, err := WriteToFile(d, dst, defFileName)
+	if err != nil {
+		return "", fmt.Errorf("WriteSrcFiletoDst failed err=%v", err)
+	}
+
+	Debug("WriteSrcFileToDst exit w success src=%s written to target=%s", src, target)
+	return target, nil
+}
+
+/*
+ * NOTE: Caller's responsbility to unmount this..use return var mountPath to unmount in caller.
+ *
+ * obtains absolute file path based on input entered by user in policy file.
+ * inputVal is of format <block device identifier>:<path>
+ * inputVal e.g sda:/boot/foo.go , 4qccd342-12zr-4e99-9ze7-1234cb1234c4:/bar/zyx.go
+ * rw_option = true --> mount is of type read-write, else read-only.
+ *
+ * This function
+ * parses user input.
+ * mounts device
+ * Get absolute file path on mounted device.
+ * does NOT Unmount device, use returned mountPath to unmount...
+ * returns filePath, mountPath, error in that order
+ */
+func GetMountedFilePath(inputVal string, rw_option bool) (string, string, error) {
+	s := strings.Split(inputVal, ":")
+	if len(s) != 2 {
+		return "", "", fmt.Errorf("%s: Usage: <block device identifier>:<path>", inputVal)
+	}
+
+	// s[0] can be sda or UUID. if UUID, then we need to find its name
+	var deviceId string = s[0]
+	if !strings.HasPrefix(deviceId, "sd") {
+		if e := getBlkInfo(); e != nil {
+			return "", "", fmt.Errorf("GetMountedFilePath: getBlkInfo err=%s", e)
+		}
+		devices := storage.PartitionsByFsUUID(storageBlkDevices, s[0]) // []BlockDev
+		for _, device := range devices {
+			Debug("device =%s with fsuuid=%s", device.Name, s[0])
+			deviceId = device.Name
+		}
+	}
+
+	devicePath := filepath.Join("/dev", deviceId) // assumes deviceId is sda, devicePath=/dev/sda
+	Debug("Attempting to mount %s", devicePath)
+	var dev *diskboot.Device
+	var err error
+	if rw_option {
+		dev, err = diskboot.FindDeviceRW(devicePath) // FindDevice fn mounts , w rw option, devicePath=/dev/sda.
+	} else {
+		dev, err = diskboot.FindDevice(devicePath) // FindDevice fn mounts devicePath=/dev/sda.
+	}
+	if err != nil {
+		return "", "", fmt.Errorf("Mount %v failed, err=%v", devicePath, err)
+	}
+
+	Debug("Mounted %s", devicePath)
+	fPath := filepath.Join(dev.MountPath, s[1]) // mountPath=/tmp/path/to/target/file if /dev/sda mounted on /tmp
+	return fPath, dev.MountPath, nil
+}
+
+/*
+ * ScanKernelCmdLine() ([]byte, error)
+ * format sl_policy=<block device identifier>:<path>
+ * e.g sda:/boot/securelaunch.policy
+ * e.g 4qccd342-12zr-4e99-9ze7-1234cb1234c4:/foo/securelaunch.policy
+ */
+func ScanKernelCmdLine() ([]byte, error) {
+
+	Debug("ScanKernelCmdLine: scanning kernel cmd line for *sl_policy* flag")
+	val, ok := cmdline.Flag("sl_policy")
+	if !ok {
+		log.Printf("ScanKernelCmdLine: sl_policy cmdline flag is not set")
+		return nil, errors.New("Flag Not Set")
```

Comment:
be more descriptive, and return it as an error rather than logging?

```go
return fmt.Errorf("cannot locate secure launch policy: sl_policy cmdline flag is not set")
```

Reply:
if i say "cannot locate secure launch policy" , user might think we stopped trying to look for it.
After this error is returned , we try scanning block devices...Here is how the error looks like on qemu with debug flag set.
`
2019/12/03 21:54:55 debug flag is set. Logging Enabled.
2019/12/03 21:54:55 ********Step 1: init completed. starting main ********
2019/12/03 21:54:55 ********Step 2: locate SL Policy ********
2019/12/03 21:54:55 ScanKernelCmdLine: scanning kernel cmd line for *sl_policy* flag
2019/12/03 21:54:55 ScanKernelCmdLine: sl_policy cmdline flag is not set
2019/12/03 21:54:55 Searching and mounting block devices with bootable configs
2019/12/03 21:54:55 scanning for policy file under devicePath=/dev/sda1, mountPath=/tmp/boot-506336525
2019/12/03 21:54:55 scanBlockDevice
2019/12/03 21:54:55 policy file found on mountPath=/tmp/boot-506336525, directory =/
2019/12/03 21:54:55 policy file found at devicePath=/dev/sda1
2019/12/03 21:54:55 policy file located
`

Reply:
maybe "cannot locate secure launch policy via kernel cmdline"? 

and the caller can log it if wanted or not?

Reply:
I guess in general I'm trying to reduce the amount of stuff logged by functions, and move it as far up in the call chain as possible -- just to keep the packages as reusable as possible by other people. 

If you think that's not attainable, feel free to ignore me here.

Reply:
Well I have always used this logic which is 
every error in a function is logged as log.Printf("Fn: err=%v", err)
and if the error is returned back 
err := fn A
if err != nil
return fmt.Errorf("A err = %v", err)
That way error call chain has a consistency...